std/crypto/sha3.zig

```diff
@@ -0,0 +1,281 @@
+    rate: usize,
+
+    pub fn init() -> Self {
+        var d: Self = undefined;
+        d.reset();
+        return d;
+    }
+
+    pub fn reset(d: &Self) {
+        mem.set(u8, d.s[0..], 0);
+        d.offset = 0;
+        d.rate = 200 - (bits / 4);
+    }
+
+    pub fn hash(b: []const u8, out: []u8) {
+        var d = Self.init();
+        d.update(b);
+        d.final(out);
+    }
+
+    pub fn update(d: &Self, b: []const u8) {
+        var ip: usize = 0;
+        var len = b.len;
+        var rate = d.rate - d.offset;
+        var offset = d.offset;
+
+        // absorb
+        while (len >= rate) {
+            for (d.s[offset .. offset + rate]) |*r, i|
+                *r ^= b[ip..][i];
+
+            keccak_f(1600, d.s[0..]);
+
+            ip += rate;
+            len -= rate;
+            rate = d.rate;
+            offset = 0;
+        }
+
+        for (d.s[offset .. offset + len]) |*r, i|
+            *r ^= b[ip..][i];
+
+        d.offset = offset + len;
+    }
+
+    pub fn final(d: &Self, out: []u8) {
+        // padding
+        d.s[d.offset] ^= delim;
+        d.s[d.rate - 1] ^= 0x80;
+
+        keccak_f(1600, d.s[0..]);
+
+        // squeeze
+        var op: usize = 0;
+        var len: usize = bits / 8;
+
+        while (len >= d.rate) {
+            mem.copy(u8, out[op..], d.s[0..d.rate]);
+            keccak_f(1600, d.s[0..]);
+            op += d.rate;
+            len -= d.rate;
+        }
+
+        mem.copy(u8, out[op..], d.s[0..len]);
+    }
+};}
+
+const RC = []const u64 {
+    0x0000000000000001, 0x0000000000008082, 0x800000000000808a, 0x8000000080008000,
+    0x000000000000808b, 0x0000000080000001, 0x8000000080008081, 0x8000000000008009,
+    0x000000000000008a, 0x0000000000000088, 0x0000000080008009, 0x000000008000000a,
+    0x000000008000808b, 0x800000000000008b, 0x8000000000008089, 0x8000000000008003,
+    0x8000000000008002, 0x8000000000000080, 0x000000000000800a, 0x800000008000000a,
+    0x8000000080008081, 0x8000000000008080, 0x0000000080000001, 0x8000000080008008,
+};
+
+const ROTC = []const usize {
+    1, 3, 6, 10, 15, 21, 28, 36,
+    45, 55, 2, 14, 27, 41, 56, 8,
+    25, 43, 62, 18, 39, 61, 20, 44
+};
+
+const PIL = []const usize {
+    10, 7, 11, 17, 18, 3, 5, 16,
+    8, 21, 24, 4, 15, 23, 19, 13,
+    12, 2, 20, 14, 22, 9, 6, 1
+};
+
+const M5 = []const usize {
+    0, 1, 2, 3, 4, 0, 1, 2, 3, 4
+};
+
+fn keccak_f(comptime F: usize, d: []u8) {
+    debug.assert(d.len == F / 8);
+
+    const B = F / 25;
+    const no_rounds = comptime x: { break :x 12 + 2 * math.log2(B); };
+
+    var s = []const u64 {0} ** 25;
+    var t = []const u64 {0} ** 1;
+    var c = []const u64 {0} ** 5;
+
+    for (s) |*r, i| {
+        *r = mem.readIntLE(u64, d[8*i .. 8*i + 8]);
+    }
+
+    var x: usize = 0;
+    var y: usize = 0;
+    // TODO: Cannot unroll all loops here due to comptime differences.
+    inline for (RC[0..no_rounds]) |round| {
+        // theta
+        x = 0; while (x < 5) : (x += 1) {
```

Comment:
I haven't been able to inline statements like these without getting wrong results in the tests. Is there any open issues relating to this or am I missing something? Or just a bug?

For example, try making making `y` `comptime` and inline all the whiles results in bad results.

Reply:
It's probably a bug. I'm going to bed now but I'll have a look tomorrow. I'll prioritize this bug since it's a silently invalid code bug.